### PR TITLE
Foreach gradient clipping

### PR DESCRIPTION
**Pull request by milesial**
Faster gradient clipping using the foreach functions

```
[------------------------ (tensors, scalar) -------------------------]
                                   |  without foreach  |  with foreach |    apex 
1 threads: ----------------------------------------------------------------------
      10 tensors of size 4         |         120.5     |       61.1    |     50.3
      100 tensors of size 4        |         946.2     |      239.5    |    136.3
      1000 tensors of size 4       |        9808.5     |     2151.1    |   1006.9
      10000 tensors of size 4      |       96871.2     |    22637.4    |  10119.1
      10 tensors of size 16        |         121.0     |       64.1    |     52.5
      100 tensors of size 16       |         993.4     |      252.6    |    136.7
      1000 tensors of size 16      |        9427.7     |     2151.2    |   1049.5
      10000 tensors of size 16     |       97437.1     |    22203.1    |  10340.0
      10 tensors of size 256       |         118.9     |       62.3    |     51.5
      100 tensors of size 256      |         955.2     |      243.1    |    134.2
      1000 tensors of size 256     |        9374.9     |     2140.7    |   1009.6
      10000 tensors of size 256    |       95302.5     |    21849.4    |  10215.5
      10 tensors of size 65536     |         118.5     |       62.4    |     51.1
      100 tensors of size 65536    |        1740.7     |      243.3    |    225.3
      1000 tensors of size 65536   |       17364.1     |     2228.7    |   2004.5
      10000 tensors of size 65536  |      177510.1     |    25410.4    |  20678.2
```

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
I'll let @janeyx99 take a look at this.
She is building a generic tool for doing this `per_device_and_dtype_grads` collection that will simplify this code.

**Review by janeyx99 — COMMENTED** [review]
Hey! Can you add a test near test_clip_grad_norm in test/test_nn.py to ensure the calculations are the same?

File: torch/nn/utils/clip_grad.py
Comment (janeyx99) [inline torch/nn/utils/clip_grad.py:71]:
```suggestion
        for grads in per_device_and_dtype_grads.values():
```
if device isn't used in this for loop

**Comment by janeyx99** [comment]
Regarding work on consolidating a util for creating this dictionary: I'm currently landing https://github.com/pytorch/pytorch/pull/92014, which has a version of this grouping function. It would be best if the functionality used across this PR could be abstracted to a common util in that file too!

File: torch/nn/utils/clip_grad.py
Comment (milesial) [reply torch/nn/utils/clip_grad.py]: Thx, changed

**Comment by milesial** [comment]
I added tests and used `_group_tensors_by_device_and_dtype`, let's see what CI says

**Comment by milesial** [comment]
@janeyx99 CI is green. I tweaked the import in your util file to avoid import race issues.

File: test/test_nn.py
Comment (janeyx99) [inline test/test_nn.py:1662]: We should probably skip the CPU when TEST_CUDA is True to not replicate this test across our CPU + GPU CI configs.

File: torch/nn/utils/clip_grad.py
Comment (janeyx99) [inline torch/nn/utils/clip_grad.py:29]: Not necessarily a thing to change on this PR, but is there any reason to not have this be the default? Is it because foreach is usually slower on CPU as it falls back to the slow path?

File: torch/nn/utils/clip_grad.py
Comment (janeyx99) [inline torch/nn/utils/clip_grad.py:68]: How come this needs to be a scalar?

**Review by janeyx99 — COMMENTED** [review]
Thanks for the fast turnaround--looks awesome overall!

I had some nits and noob questions.

**Comment by janeyx99** [comment]
I confirm the tests passed! 
image

File: test/test_nn.py
Comment (milesial) [reply test/test_nn.py:1662]: done

File: torch/nn/utils/clip_grad.py
Comment (milesial) [reply torch/nn/utils/clip_grad.py]: Doesn't need to actually.
Just fixed in new commit to avoid a sync

Comment (milesial) [reply torch/nn/utils/clip_grad.py]: Ah I remember why I did it, it works fine by mypy complains
```
No overload variant of "_foreach_mul_" matches argument types
    "List[Tensor]", "Tensor"
```

File: torch/nn/utils/clip_grad.py
Comment (janeyx99) [reply torch/nn/utils/clip_grad.py:29]: Hahaha I am currently working on making foreach the default for optimizers, but there's a bit more complexity there since we have an option for differentiable. Likewise, there are other backends like mps and rocm that don't support foreach.

I think a good plan is to make the foreach kwarg an Optional[bool], to give the user ability to choose True or False, and then for us, since we know it's definitely faster on cpu and on cuda to turn on foreach = True if foreach was not specified by the user + the tensors are on CPU or CUDA.

File: torch/nn/utils/clip_grad.py
Comment (janeyx99) [inline torch/nn/utils/clip_grad.py:69]: Oh I see why you had it as an item call before..._foreach_mul either takes in a tensorlist + scalar or a tensorlist and a tensorlist.

Sorry for not seeing this earlier!

Comment (milesial) [reply torch/nn/utils/clip_grad.py:69]: But this way works too! Since somewhere the 0-dim tensor is seen as a scalar. And it should avoid a sync. But mypy is not happy with that, not sure how to fix it, see the failing test.

File: torch/nn/utils/clip_grad.py
Comment (milesial) [reply torch/nn/utils/clip_grad.py:29]: Yes makes sense, I'll turn it on. But why Optional[bool] and not just a bool with default True?

Comment (milesial) [reply torch/nn/utils/clip_grad.py:29]: If tensors are on a non CPU/CUDA device, won't the foreach call just go to "slow" path (which would still be faster than python)? If so there isn't any issue with setting it to default True regardless of device. Can't test it on a MPS device on my side.

File: test/test_nn.py
Comment (janeyx99) [inline test/test_nn.py:1658]: Another nit! We already have perdevice test classes where device will be auto-parametrized (it is provided as an argument to the test function). See `test_affine_2d_rotate90` as an example. You can move these tests under NNDeviceType if they aren't already and we can remove the logic where we have to skip based on device_type.

File: torch/nn/utils/clip_grad.py
Comment (milesial) [reply torch/nn/utils/clip_grad.py:29]: Also since we split the tensors by device anyway we could run _foreach only on the cpu/cuda tensor groups and do the python for loop for the rest

File: torch/nn/utils/clip_grad.py
Comment (janeyx99) [reply torch/nn/utils/clip_grad.py]: Oh, in that case...
```suggestion
            torch._foreach_mul_(grads, clip_coef_clamped.to(grads[0].device))  # type: ignore[call-overload]
```
🙃

File: torch/nn/utils/clip_grad.py
Comment (milesial) [reply torch/nn/utils/clip_grad.py]: Since we split the tensors by device anyway we could run _foreach only on the cpu/cuda tensor groups and do the python for loop for the rest. See my commit https://github.com/pytorch/pytorch/pull/91846/commits/618a54a47aa53ccc65599ab0c292a66d04f46da4 . This commit also makes the non-foreach version slightly faster thanks to the per-device grouping.

File: test/test_nn.py
Comment (milesial) [reply test/test_nn.py]: Done in https://github.com/pytorch/pytorch/pull/91846/commits/adde43969b2e0fb04e313660bf6373d7879e6db5 but now the tests don't run when I locally call pytest

File: torch/nn/utils/clip_grad.py
Comment (janeyx99) [reply torch/nn/utils/clip_grad.py]: Also, is the sending to the device necessary for the scalar tensor? Could you try without the device call?

Comment (milesial) [reply torch/nn/utils/clip_grad.py]: Yes I figured it out too thanks!

File: test/test_nn.py
Comment (janeyx99) [inline test/test_nn.py:11412]: Not a concern with your PR, but I am realizing we never run this in CI because we only have one CI config where there is more than one GPU and we don't run this test in that config. 🤔 Filed https://github.com/pytorch/pytorch/issues/92173

File: torch/nn/utils/clip_grad.py
Comment (janeyx99) [inline torch/nn/utils/clip_grad.py:14]: So we actually cannot just set it to True, because most foreach ops only have a dispatch defined for CUDA and CPU (and not any other backend like MKLDNN or MPS etc) (see the ops in native_functions.yaml)

File: torch/utils/_foreach_utils.py
Comment (janeyx99) [inline torch/utils/_foreach_utils.py:22]: thanks for the cleanup here!

File: torch/nn/utils/clip_grad.py
Comment (janeyx99) [reply torch/nn/utils/clip_grad.py]: noice!

File: test/test_nn.py
Comment (janeyx99) [reply test/test_nn.py]: huh i don't use pytest but weirdddd @clee2000 do you know how to run specific tests in pytest once they're subscribed to a device type class?

File: test/test_nn.py
Comment (janeyx99) [reply test/test_nn.py:11412]: If you want to see this run in CI, you would have to add a line similar to https://github.com/pytorch/pytorch/pull/92048/files#diff-5356a2d45f3d28e01b954926d7f1681cccb0c9ad2cafd722478de8c283090bd3R48

Comment (janeyx99) [reply test/test_nn.py:11412]: And then you would also need to add the ciflow/periodic label to get the multigpu tests to trigger.

File: torch/nn/utils/clip_grad.py
Comment (janeyx99) [inline torch/nn/utils/clip_grad.py:70]:
```suggestion
            torch._foreach_mul_(grads, clip_coef_clamped.to(device))  # type: ignore[call-overload]
```

File: test/test_nn.py
Comment (janeyx99) [inline test/test_nn.py:16]:
```suggestion
```

File: torch/nn/utils/clip_grad.py
Comment (milesial) [reply torch/nn/utils/clip_grad.py:14]: In the newest commit the behaviour is that if foreach is True, it will run foreach on CPU and CUDA, but still keep running the other device groups like before with a python for loop. That's okay right?

File: torch/utils/_foreach_utils.py
Comment (milesial) [reply torch/utils/_foreach_utils.py:22]: np, I hope assuming that the device type will keep being hashable is ok

File: test/test_nn.py
Comment (milesial) [reply test/test_nn.py:11411]: ok my bad they actually run as expected with `pytest -k clip_grad test/test_nn.py`, not sure what I did wrong before

File: test/test_nn.py
Comment (milesial) [reply test/test_nn.py]: ok my bad they actually run as expected with `pytest -k clip_grad test/test_nn.py`, not sure what I did wrong before

File: torch/nn/utils/clip_grad.py
Comment (janeyx99) [reply torch/nn/utils/clip_grad.py:14]: Yes, that's right. cc @albanD to confirm we aren't falling subject to BC breaking risks by having foreach to default to True here.

Could you please document the default for the foreach kwarg in both the function documentations?

Comment (albanD) [reply torch/nn/utils/clip_grad.py:14]: I think one of the benefit to have 3 states (None, True, False) is that the user has an API to force this to be used and they will get a nice error if it doesn't.
I think it is important to have such a nob for users who care about perf.

Comment (milesial) [reply torch/nn/utils/clip_grad.py]: Got it, I implemented the 3-way state and doc changes in https://github.com/pytorch/pytorch/pull/91846/commits/cfea3967a017befad38829e75cbc95fb966261f5

**Review by janeyx99 — APPROVED** [review]
Looks good to me! Thank you very much for the perf speedup :D 

(sorry about the merge conflict--that's my bad)

**Comment by milesial** [comment]
I rebased but the signature change is breaking torch XLA since the patching there expects the old signature

```
 Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/xla/test/../../test/test_view_ops.py", line 15, in <module>
    from torch.testing._internal.common_device_type import \
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 603, in <module>
    mod = runpy.run_path(path, init_globals=globals())  # type: ignore[func-returns-value]
  File "/opt/conda/lib/python3.7/runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "/opt/conda/lib/python3.7/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/opt/conda/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/var/lib/jenkins/workspace/xla/test/pytorch_test_base.py", line 8, in <module>
    import torch_xla
  File "/opt/conda/lib/python3.7/site-packages/torch_xla-2.0.0-py3.7-linux-x86_64.egg/torch_xla/__init__.py", line 140, in <module>
    _apply_patches()
  File "/opt/conda/lib/python3.7/site-packages/torch_xla-2.0.0-py3.7-linux-x86_64.egg/torch_xla/_patched_functions.py", line 54, in _apply_patches
    nn.utils.clip_grad_norm_ = _patch(nn.utils.clip_grad_norm_, clip_grad_norm_)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla-2.0.0-py3.7-linux-x86_64.egg/torch_xla/_patched_functions.py", line 16, in _patch
    fn, xfingerprint, fingerprint))
RuntimeError: Unable to patch <function clip_grad_norm_ at 0x7fc8b282d950>, signature mismatch: (parameters: Union[torch.Tensor, Iterable[torch.Tensor]], max_norm: float, norm_type: float = 2.0, error_if_nonfinite: bool = False, foreach: bool = None) -> torch.Tensor vs (parameters: Union[torch.Tensor, Iterable[torch.Tensor]], max_norm: float, norm_type: float = 2.0, error_if_nonfinite: bool = False) -> torch.Tensor
```

from https://github.com/pytorch/xla/blob/d636e7774b63cc070d7ebbfeec950e4892efa713/torch_xla/_patched_functions.py#L21-L54

How should we approach that, first merge a fix to torch xla to allow both old and new signatures, then merge this PR?

**Comment by janeyx99** [comment]
Yes, we'd want to sync the lands. If we're able to fix xla without breaking pytorch, then go for it and land a patch there first. In general, we follow these steps, but the force merging may be unnecessary if xla can be green the whole time:

(1) Make a pytorch/pytorch PR and a pytorch/xla patch
(2) update xla.txt on the pytorch.pytorch PR to point to your patch
(3) once pytorch/pytorch CI is *fully green*, rebase on tip-of-master again to minimize the chance of merge conflicts / last minute problems
(4) once you get fully green CI on the newly rebased pytorch/pytorch PR, merge the pytorch/xla PR (XLA CI will start failing)
(5) update the pytorch/pytorch PR to the new tip-of-master XLA commit hash (no other changes should be required), and immediately force-merge. You’re counting on the fact that CI was green ~3 hours ago.

The force merge is because we’re betting on the fact that nothing should have changed from the last run to the next, and we don’t want to keep XLA CI red for an unnecessary 3 hours. And the rebasing beforehand is because at least so far, merge conflicts have been a frequent source of “the pytorch/xla PR merged, but the pytorch/pytorch PR is no longer ready”

An example can be found by following https://github.com/pytorch/xla/blob/d636e7774b63cc070d7ebbfeec950e4892efa713/.circleci/README.md?plain=1#L10

**Comment by milesial** [comment]
I don't think I can do it from 2) since I don't have write access to the xla repo, I only have a fork and xla.txt can't be in a fork I think. I tried setting it to `pull/4471/head` but that didn't work

**Comment by janeyx99** [comment]
looping in XLA folks @JackCaoG could you help with the XLA side of the pinning?

**Comment by JackCaoG** [comment]
@wonjoolee95 can you take a look? I think there is a companion pr opened on xla side, we just need to do the pinging.

**Comment by wonjoolee95** [comment]
I've opened opened https://github.com/pytorch/xla/pull/4480 (picking up @milesial's https://github.com/pytorch/xla/pull/4471). 

@milesial, can we update the XLA's commit pin (https://github.com/pytorch/pytorch/blob/master/.github/ci_commit_pins/xla.txt) on this PR to `eddaa4b3cf7c4c9302b6b04c6e5d13b4c6ba260b` and let the CI verify? Thanks!

**Comment by milesial** [comment]
Thanks, just did. Let's see what CI thinks of it

**Comment by wonjoolee95** [comment]
So, it looks like the XLA's PR is treating the `RuntimeError` as failures so updated the https://github.com/pytorch/xla/pull/4480 to explicitly skip the `test_clip_grad_value_foreach_True_*` and `test_clip_grad_norm_foreach_True_*` tests.

**Comment by wonjoolee95** [comment]
Looks like the CI is green on both sides. Let's coordinate a merge tomorrow.

**Comment by milesial** [comment]
@wonjoolee95 I rebased this MR. Once it's green you can merge the XLA MR. Then I'll update the pin on this MR and force merge.

**Comment by milesial** [comment]
@wonjoolee95 CI passed, can you merge the XLA PR?

**Comment by wonjoolee95** [comment]
> @wonjoolee95 CI passed, can you merge the XLA PR?

@milesial, just merged to master. The new pin should be `eac4e547138ab22a9b41c6f96208613fd7dd19d5`.

**Comment by milesial** [comment]
Thanks, let's hope it goes smoothly

@pytorchbot merge -f "coordinating merge with XLA, CI passed"

**Comment by JackCaoG** [comment]
FYI, next time we can merge this pr first(pin can point to a unmerged branch and that's by design), and then merge the xla one. Otherwise xla head will be red until this pr merged.

**Comment by milesial** [comment]
Haha I can't force merge, @janeyx99 can you help?

**Comment by wonjoolee95** [comment]
Okay since we can't force merge this right now, I'm going to revert the XLA's PR lol.

**Comment by JackCaoG** [comment]
@milesial don't worry about our revert, as long as pytorch still pin to the correct pytorch/xla pin, pytorch CI will be fine.

**Comment by wonjoolee95** [comment]
You can update the XLA pin in this PR to `8dcab83819368f468dadbe6e81b064d268830df2` and `merge -g`. I'll merge the XLA's companion PR once this merges.

**Comment by milesial** [comment]
I think keeping `eac4e` is fine no? No need to switch to `8dca` right?

@pytorchbot merge -g

**Comment by JackCaoG** [comment]
yea, you can keep the old pin

**Comment by janeyx99** [comment]
Oh I can force merge :D

**Comment by janeyx99** [comment]
don't speak too soon 🙃

**Comment by ngimel** [comment]
@milesial @crcrpar can you check if debug build using this op errors out? We have reports of debug builds erroring out with 
```
** RuntimeError: t.storage().use_count() == 1 INTERNAL ASSERT FAILED at "caffe2/torch/csrc/autograd/autograd_not_implemented_fallback.cpp":189, please report a bug to PyTorch.
```

**Comment by milesial** [comment]
I'll check.
By debug build you mean building with `DEBUG=1` right?

**Pull request closed without merging**
